src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs

```diff
@@ -1597,6 +1598,87 @@ public static IEnumerable<IEdmStructuralProperty> Key(this IEdmEntityType type)
+                if (declaredAlternateKeys != null)
+                {
+                    return declaredAlternateKeys;
+                }
+
+                checkingType = checkingType.BaseEntityType();
+            }
+
+            return Enumerable.Empty<IDictionary<string, IEdmProperty>>();
+        }
+
+        /// <summary>
+        /// Adds the key to the alternateKeys of this entity type.
+        /// </summary>
+        /// <param name="type">Reference to the calling object.</param>
+        /// <param name="model">The model to be used.</param>
+        /// <param name="alternateKey">Dictionary of alias and structural properties for the alternate key.</param>
+        public static void AddAlternateKey(this IEdmEntityType type, EdmModel model, IDictionary<string, IEdmProperty> alternateKey)
+        {
+            EdmUtil.CheckArgumentNull(type, "type");
+            EdmUtil.CheckArgumentNull(model, "model");
+            EdmUtil.CheckArgumentNull(alternateKey, "alternateKey");
+
+            EdmCollectionExpression annotationValue = null;
+            var ann = model.FindVocabularyAnnotations<IEdmValueAnnotation>(type, AlternateKeysVocabularyModel.AlternateKeysTerm).FirstOrDefault();
+            if (ann != null)
+            {
+                annotationValue = ann.Value as EdmCollectionExpression;
+            }
+
+            List<IEdmExpression> alternateKeysCollection;
+            if (annotationValue != null)
+            {
+                alternateKeysCollection = new List<IEdmExpression>(annotationValue.Elements);
+            }
+            else
+            {
+                alternateKeysCollection = new List<IEdmExpression>();
+            }
+
+            List<IEdmExpression> propertyRefs = new List<IEdmExpression>();
+
+            foreach (KeyValuePair<string, IEdmProperty> kvp in alternateKey)
+            {
+                IEdmRecordExpression propertyRef = new EdmRecordExpression(
+                    new EdmComplexTypeReference(AlternateKeysVocabularyModel.PropertyRefType, false), 
+                    new EdmPropertyConstructor(AlternateKeysVocabularyConstants.PropertyRefTypeAliasPropertyName, new EdmStringConstant(kvp.Key)),
+                    new EdmPropertyConstructor(AlternateKeysVocabularyConstants.PropertyRefTypeNamePropertyName, new EdmPropertyPathExpression(kvp.Value.Name)));
+                propertyRefs.Add(propertyRef);
+            }
+
+            EdmRecordExpression alternateKeyRecord = new EdmRecordExpression(
+                new EdmComplexTypeReference(AlternateKeysVocabularyModel.AlternateKeyType, false),
+                new EdmPropertyConstructor(AlternateKeysVocabularyConstants.AlternateKeyTypeKeyPropertyName, new EdmCollectionExpression(propertyRefs)));
+
+            alternateKeysCollection.Add(alternateKeyRecord);
+
+            var annotation = new EdmAnnotation(

```

Comment:
In case of existing annots we reinitialize using alternateKeysCollection = new List<IEdmExpression>(annotationValue.Elements); 

We cannot add since (annotationValue.Value as IEdmCollectionExpression).Elements is an IEnumberable. Cannot add to it.